### PR TITLE
Fix lora single device fine tune checkpoint saving & nan loss when use_dora=True

### LINKED ISSUE #1903: [bug] DoRA is broken
Two separate DoRA bugs I just noticed:

(1) Llama 3.2 1B config with DoRA errors on state dict load. Repro:

```
tune run lora_finetune_single_device --config llama3_2/1B_lora_single_device \
gradient_accumulation_steps=1 max_steps_per_epoch=5 model.use_dora=True
...
Exception: Error converting the state dict. Found unexpected key: "layers.0.attn.q_proj.magnitude". Please make sure you're loading a checkpoint with the right format.
```

(2) Llama 3.2 Vision 11B model with DoRA has NaN loss. Repro:

```
tune run lora_finetune_single_device --config llama3_2_vision/11B_lora_single_device \
max_steps_per_epoch=5 gradient_accumulation_steps=1 model.use_dora=True
```

Once we fix them we should add recipe test cases setting `model.use_dora=True` to catch these errors in the future, cc @felipemello1.

### PR DESCRIPTION
**Pull request by mirceamironenco**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Fixes #1903 .

#### Changelog
What are the changes made in this PR?
* Account for the magnitude parameter when constructing the state mapping dict (hf/peft).
* Initialize dora magnitude in the single device lora finetune recipe.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [x] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: torchtune/models/convert_weights.py [hunk 1]
```diff
@@ -260,15 +260,21 @@ def tune_to_peft_adapter_weights(
...
+                        )
+                    }
+                )
+            else:
+                full_mapping.update(
+                    {vv.replace(".weight", f".{k}"): kk.replace(".weight", f".{v}")}
                 )
-                for kk, vv in _FROM_HF.items()
-                if vv is not None
-            }
```
Comment (SalmanMohammadi) [inline torchtune/models/convert_weights.py:270]: Is this just for catching the DoRA magnitude parameter which should map to `{layer}.magnitude` rather than `{layer}.magnitude.weight`? Or are there any other cases to catch? 
If so why not change the control flow to explictly catch this case so it's super clear why we need different logic. If not, ignore me : )

Comment (mirceamironenco) [reply torchtune/models/convert_weights.py:270]: Yes that's correct. You mean changing it to something like `if k == "magnitude": (second branch of current code) else: (first branch of current code)`? Or do you want me to also modify the specific form of the dict item (`{vv.replace(".weight", f".{k}"): kk.replace(".weight", f".{v}")}`)?

Comment (SalmanMohammadi) [reply torchtune/models/convert_weights.py:270]: > You mean changing it to something like if k == "magnitude": (second branch of current code) else: (first branch of current code)

I think this sounds good! If you wanted to be clever about it you could do something like
```python

**Comment by SalmanMohammadi** [comment]
Thank you so much for contributing this fix @mirceamironenco. Would you be able to share a small training run showing that this works OK? 

Another thing that would be great to see would be modifying one of our LoRA recipe tests to make sure the SD loading logic works as expected for DoRA. You could add this here https://github.com/pytorch/torchtune/blob/09c2619f713e771b4159f7b83bac8971c7053bd3/tests/recipes/test_lora_finetune_single_device.py#

but I won't block on this - happy to file an issue for a follow up.

**Comment by mirceamironenco** [comment]
> Thank you so much for contributing this fix @mirceamironenco. Would you be able to share a small training run showing that this works OK?

Running `tune run lora_finetune_single_device --config llama3_2/1B_lora_single_device \
gradient_accumulation_steps=1 max_steps_per_epoch=5 model.use_dora=True`:

```
(config here)
DEBUG:torchtune.utils._logging:Setting manual seed to local seed 3727881134. Local seed is seed + rank = 3727881134 + 0
Writing logs to /tmp/lora_finetune_output/log_1730294718.txt
INFO:torchtune.utils._logging:Model is initialized with precision torch.bfloat16.
INFO:torchtune.utils._logging:Memory stats after model init:
        GPU peak memory allocation: 2.52 GiB
        GPU peak memory reserved: 2.56 GiB
        GPU peak memory active: 2.52 GiB
INFO:torchtune.utils._logging:Tokenizer is initialized from file.
INFO:torchtune.utils._logging:Optimizer and loss are initialized.
INFO:torchtune.utils._logging:Loss is initialized.
INFO:torchtune.utils._logging:Dataset and Sampler are initialized.
INFO:torchtune.utils._logging:Learning rate scheduler is initialized.
WARNING:torchtune.utils._logging: Profiling disabled.
INFO:torchtune.utils._logging: Profiler config after instantiation: {'enabled': False}
1|5|Loss: 1.7691251039505005: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [00:01<00:00,  2.80it/s]
INFO:torchtune.utils._logging:Starting checkpoint save...
INFO:torchtune.utils._logging:Model checkpoint of size 2.47 GB saved to /tmp/Llama-3.2-1B-Instruct/hf_model_0001_0.pt
INFO:torchtune.utils._logging:Adapter checkpoint of size 0.09 GB saved to /tmp/Llama-3.2-1B-Instruct/adapter_0.pt
INFO:torchtune.utils._logging:Adapter checkpoint of size 0.09 GB saved to /tmp/Llama-3.2-1B-Instruct/adapter_model.bin
INFO:torchtune.utils._logging:Adapter checkpoint of size 0.00 GB saved to /tmp/Llama-3.2-1B-Instruct/adapter_config.json
INFO:torchtune.utils._logging:Saving final epoch checkpoint.
INFO:torchtune.utils._logging:The full model checkpoint, including all weights and configurations, has been saved successfully.You can now use this checkpoint for further training or inference.
INFO:torchtune.utils._logging:Checkpoint saved in 222.20 seconds.
1|5|Loss: 1.7691251039505005: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 5/5 [03:43<00:00, 44.78s/it]
```
Log file:
```
Step 1 | loss:2.0422861576080322 lr:2.9999999999999997e-06 tokens_per_second_per_gpu:1470.76318359375 peak_memory_active:5.774061679840088 peak_memory_alloc:5.774061679840088 peak_memory_reserved:6.09765625
Step 2 | loss:2.2708561420440674 lr:5.999999999999999e-06 tokens_per_second_per_gpu:2653.373291015625 peak_memory_active:5.9765753746032715 peak_memory_alloc:5.9765753746032715 peak_memory_reserved:6.216796875
Step 3 | loss:1.9193693399429321 lr:8.999999999999999e-06 tokens_per_second_per_gpu:2741.960693359375 peak_memory_active:7.017248630523682 peak_memory_alloc:7.017248630523682 peak_memory_reserved:7.466796875
Step 4 | loss:1.6952054500579834 lr:1.1999999999999999e-05 tokens_per_second_per_gpu:2495.660400390625 peak_memory_active:7.290591716766357 peak_memory_alloc:7.290591716766357 peak_memory_reserved:7.84765625
Step 5 | loss:1.7691251039505005 lr:1.4999999999999999e-05 tokens_per_second_per_gpu:2274.15478515625 peak_memory_active:10.266940593719482 peak_memory_alloc:10.266940593719482 peak_memory_reserved:10.51171875

```

**Comment by SalmanMohammadi** [comment]
> ```
> Step 1 | loss:2.0422861576080322 lr:2.9999999999999997e-06 tokens_per_second_per_gpu:1470.76318359375 peak_memory_active:5.774061679840088 peak_memory_alloc:5.774061679840088 peak_memory_reserved:6.09765625
> Step 2 | loss:2.2708561420440674 lr:5.999999999999999e-06 tokens_per_second_per_gpu:2653.373291015625 peak_memory_active:5.9765753746032715 peak_memory_alloc:5.9765753746032715 peak_memory_reserved:6.216796875
> Step 3 | loss:1.9193693399429321 lr:8.999999999999999e-06 tokens_per_second_per_gpu:2741.960693359375 peak_memory_active:7.017248630523682 peak_memory_alloc:7.017248630523682 peak_memory_reserved:7.466796875
> Step 4 | loss:1.6952054500579834 lr:1.1999999999999999e-05 tokens_per_second_per_gpu:2495.660400390625 peak_memory_active:7.290591716766357 peak_memory_alloc:7.290591716766357 peak_memory_reserved:7.84765625
> Step 5 | loss:1.7691251039505005 lr:1.4999999999999999e-05 tokens_per_second_per_gpu:2274.15478515625 peak_memory_active:10.266940593719482 peak_memory_alloc:10.266940593719482 peak_memory_reserved:10.51171875
> ```

Thanks for running this! I'll try out the vision config. 
One thing - what's up with the memory usage? 5GB going up to 10GB in 5 steps - is that normal?

**Comment by ebsmothers** [comment]
Thanks for working on this! Looks like CI is red though. Re @SalmanMohammadi's question about memory consumption, we can pull some profiles to figure out what's going on. In my experience DoRA is definitely more memory hungry, but a 2 GB difference seems high for a model of this size. It's not a blocker to land the bug fix, but I'm curious whether the increase in memory scales with model size (if so we could run into problems with 7B+ models). As for the fact that it increases over the first N steps that could be OK (and expected) given batch-dependent sequence lengths, as long as it stabilizes.

**Review by ebsmothers — APPROVED** [review]
Thanks @mirceamironenco for fixing this! This looks great to me

**Comment by ebsmothers** [comment]
(Very) belated discovery: this breaks DoRA distributed training. E.g. 

```
tune run --nnodes 1 --nproc_per_node 2 lora_finetune_distributed --config llama3/8B_dora
```

fails after this commit due to magnitude still being on meta device. 22 @SLR722
